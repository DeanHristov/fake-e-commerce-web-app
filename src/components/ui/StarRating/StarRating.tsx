'use client';

import { FC, useCallback, useState } from 'react';

export interface IStarRatingProps {
  activeColor: string;
  inActiveColor: string;
  rating: number;
  starCount?: number;
  onChange?: (num: number) => void;
}

const StarRating: FC<IStarRatingProps> = ({
  rating,
  activeColor,
  inActiveColor,
  starCount = 5,
  onChange = (num: number) => {},
}) => {
  const [activeRating, setActiveRating] = useState(rating);
  const handlerClick = useCallback(
    (vote: number) => {
      setActiveRating(vote + 1);
      onChange(vote + 1);
    },
    [onChange],
  );

  const stars: string[] = Array.from({ length: starCount }, () => `â˜†`);
  return (
    <div data-testid="star-rating" className="flex space-x-2 items-center mt-3">
      <div className="flex space-x-1 justify-start items-baseline">
        {stars.map((star, idx) => {
          let style = inActiveColor;

          if (idx < Math.round(activeRating)) style = activeColor;

          return (
            <span
              data-testid="star-item"
              style={{ color: style }}
              key={`star-${idx}`}
              className="star-icon"
              onClick={() => handlerClick(idx)}
            >
              {star}
            </span>
          );
        })}
      </div>
      <span
        data-testid="rating"
        className="text-xs font-medium text-zambezi leading-4"
      >
        {`${rating} Reviews`}
      </span>
    </div>
  );
};

export default StarRating;
