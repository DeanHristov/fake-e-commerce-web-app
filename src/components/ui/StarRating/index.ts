import StarRating, { IStarRatingProps } from './StarRating';

export type { IStarRatingProps };
export default StarRating;
