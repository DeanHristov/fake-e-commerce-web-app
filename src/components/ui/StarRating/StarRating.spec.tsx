import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import StarRating from './';

describe('UI/Component <StarRating {...} />', () => {
  it('Should render the default state', () => {
    const { container } = render(
      <StarRating
        activeColor="#1B1B1B"
        inActiveColor="#D1D1D1"
        starCount={5}
        rating={2.5}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Should be visible', () => {
    const { getByTestId } = render(
      <StarRating
        activeColor="#1B1B1B"
        inActiveColor="#D1D1D1"
        starCount={5}
        rating={2.5}
      />,
    );

    expect(getByTestId('star-rating')).toBeVisible();
  });

  it('Should be able to render exact amount of stars', () => {
    const { getAllByTestId } = render(
      <StarRating
        activeColor="#1B1B1B"
        inActiveColor="#D1D1D1"
        starCount={5}
        rating={2.5}
      />,
    );

    const starsArray: HTMLElement[] = getAllByTestId('star-item');
    expect(starsArray.length).toEqual(5);
  });

  it('Should be able to trigger click event', async () => {
    const mockCB = jest.fn();

    const { getAllByTestId } = render(
      <StarRating
        activeColor="#1B1B1B"
        inActiveColor="#D1D1D1"
        starCount={5}
        rating={2.5}
        onChange={mockCB}
      />,
    );

    const starsArray: HTMLElement[] = getAllByTestId('star-item');

    await userEvent.click(starsArray[0]);

    expect(mockCB).toHaveBeenCalledWith(1);
  });
});
