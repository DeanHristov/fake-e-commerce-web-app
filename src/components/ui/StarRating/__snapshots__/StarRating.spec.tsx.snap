// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <StarRating {...} /> Should render the default state 1`] = `
<div>
  <div
    class="flex space-x-2 items-center mt-3"
    data-testid="star-rating"
  >
    <div
      class="flex space-x-1 justify-start items-baseline"
    >
      <span
        class="star-icon"
        data-testid="star-item"
        style="color: rgb(27, 27, 27);"
      >
        ☆
      </span>
      <span
        class="star-icon"
        data-testid="star-item"
        style="color: rgb(27, 27, 27);"
      >
        ☆
      </span>
      <span
        class="star-icon"
        data-testid="star-item"
        style="color: rgb(27, 27, 27);"
      >
        ☆
      </span>
      <span
        class="star-icon"
        data-testid="star-item"
        style="color: rgb(209, 209, 209);"
      >
        ☆
      </span>
      <span
        class="star-icon"
        data-testid="star-item"
        style="color: rgb(209, 209, 209);"
      >
        ☆
      </span>
    </div>
    <span
      class="text-xs font-medium text-zambezi leading-4"
      data-testid="rating"
    >
      2.5 Reviews
    </span>
  </div>
</div>
`;
